autogpt_platform/backend/backend/server/v2/library/db.py

```diff
@@ -1,165 +1,633 @@
+        raise backend.server.v2.store.exceptions.DatabaseError(
+            "Failed to update library agent"
+        ) from e
+
+
+async def add_store_agent_to_library(
+    store_listing_version_id: str, user_id: str
+) -> Optional[prisma.models.LibraryAgent]:
+    """
+    Adds an agent from a store listing version to the user's library if they don't already have it.
+
+    Args:
+        store_listing_version_id: The ID of the store listing version containing the agent.
+        user_id: The user’s library to which the agent is being added.
+
+    Returns:
+        The newly created LibraryAgent if successfully added, or None if the agent already exists.
+
+    Raises:
+        AgentNotFoundError: If the store listing or associated agent is not found.
+        DatabaseError: If there's an issue creating the LibraryAgent record.
+    """
+    logger.debug(
+        "Adding agent from store listing version %s to library for user %s",
+        store_listing_version_id,
+        user_id,
     )
 
     try:
-        # Get store listing version to find agent
         store_listing_version = (
             await prisma.models.StoreListingVersion.prisma().find_unique(
                 where={"id": store_listing_version_id}, include={"Agent": True}
             )
         )
-
         if not store_listing_version or not store_listing_version.Agent:
             logger.warning(
-                f"Store listing version not found: {store_listing_version_id}"
+                "Store listing version not found or has no Agent: %s",
+                store_listing_version_id,
             )
             raise backend.server.v2.store.exceptions.AgentNotFoundError(
-                f"Store listing version {store_listing_version_id} not found"
+                f"Store listing version {store_listing_version_id} not found or invalid."
             )
 
         agent = store_listing_version.Agent
-
         if agent.userId == user_id:
             logger.warning(
-                f"User {user_id} cannot add their own agent to their library"
+                "User %s attempted to add their own agent to their library", user_id
             )
             raise backend.server.v2.store.exceptions.DatabaseError(
-                "Cannot add own agent to library"
+                "Cannot add own agent to library."
             )
 
         # Check if user already has this agent
-        existing_user_agent = await prisma.models.UserAgent.prisma().find_first(
+        existing_user_agent = await prisma.models.LibraryAgent.prisma().find_first(
             where={
                 "userId": user_id,
                 "agentId": agent.id,
                 "agentVersion": agent.version,
             }
         )
-
         if existing_user_agent:
-            logger.debug(
-                f"User {user_id} already has agent {agent.id} in their library"
-            )
-            return
+            logger.debug("User %s already has agent %s in library", user_id, agent.id)
```

Comment:
probs should have a notify to the user here via an exception or similar?

Reply:
- #9515